torchbiggraph/examples/e2e/data_prep.py

```diff
@@ -0,0 +1,325 @@
+    """
+    subquery = ""
+    first = True
+    for i in range(nctes):
+        if not first:
+            subquery += "\tunion\n"
+        subquery += f"\tselect * from cte_{i}\n"
+        first = False
+    return subquery
+
+
+def remap_edges(conn, rels, entity2partitions):
+    """
+    A function to remap all edges to ordinal IDs
+    according to their type.
+    """
+    logging.info("Remapping edges")
+    start = time.time()
+
+    nentities_premap = conn.execute("""
+    select count(*) from edges
+    ;
+    """).fetchall()[0][0]
+
+    query = ""
+    NPARTS = max(entity2partitions.values())
+    for lhs_part in range(NPARTS):
+        for rhs_part in range(NPARTS):
+            nctes, ctes = generate_ctes(lhs_part, rhs_part, rels, entity2partitions)
+            subquery = generate_unions(nctes)
+            query += edges_partitioned.format(
+                i = lhs_part,
+                j = rhs_part,
+                ctes=ctes,
+                tables=subquery
+            )
+
+    logging.debug(f"Running query: {query}")
+    conn.executescript(query)
+
+    logging.debug("Confirming that we didn't drop any edges.")
+    nentities_postmap = 0
+    for lhs_part in range(NPARTS):
+        for rhs_part in range(NPARTS):
+            nentities_postmap += conn.execute(f"""
+            select count(*) from edges_{lhs_part}_{rhs_part}
+            """).fetchall()[0][0]
+    
+    if nentities_postmap != nentities_premap:
+        logging.warning("DROPPED EDGES DURING REMAPPING.")
+        logging.warning(f"We started with {nentities_premap} and finished with {nentities_postmap}")
+
+    end = time.time()
+    logging.info(f"Remapped edges in {end - start}s")
+
+
+def load_edges(fname, conn):
+    """
+    A simple function to load the edges into the SQL table. It is
+    assumed that we will have a file of the form:
+    | source_id | source_type | relationship_name | destination_id | destination_type |
+    """
+    logging.info("Loading edges")
+    start = time.time()
+    cur = conn.cursor()
+    cur.executescript("""
+    DROP TABLE IF EXISTS edges
+    ;
+
+    CREATE TABLE edges (
+        source_id INTEGER,
+        source_type TEXT,
+        destination_id INTEGER,
+        destination_type TEXT,
+        rel TEXT
+    )
+    """)
+
+    edges = pd.read_csv(fname)
+    edges.to_sql('edges', conn, if_exists='append', index=False)
+    end = time.time()
+    logging.info(f"Loading edges in {end - start}s")
+
+
+def write_relations(outdir, rels, conn):
+    """
+    A simple function to write the relevant relationship information out
+    for training.
+    """
+    logging.info("Writing relations for training")
+    start = time.time()
+    out = rels.sort_values('graph_id')['id'].to_list()
+    with open(f'{outdir}/dynamic_rel_names.json', mode='w') as f:
+        json.dump(out, f, indent=4)
+    end = time.time()
+    logging.info(f"Wrote relations in {end - start}s")
+
+
+def write_single_edge(work_packet):
```

Comment:
```suggestion
def write_single_bucket(work_packet):
```

At first I thought you were doing this for every single edge :)